network_simulator/statistics.py

```diff
@@ -4,6 +4,186 @@
+        # Create a tuple of the pair and check if it exists in the dict.
+        tup = (link.end_1_addr, link.end_2_addr)
+
+        if tup not in self.link_stats:
+            self.link_stats[tup] = LinkStats
+        return self.link_stats[tup]
+
+    def get_flow_stats(self, flow):
+        """
+        Check whether a flow has already existed in the dictionary; if so,
+        skip the rest of the steps and return the stats object. If not,
+        create a new mapping then return the stats object.
+
+        :param Flow flow: flow to be determined.
+        """
+        if flow.id not in self.flow_stats:
+            self.flow_stats[flow.id] = FlowStats
+        return self.flow_stats[flow.id]
+
+    def get_host_stats(self, host):
+        """
+        Check whether a host has already existed in the dictionary; if so,
+        skip the rest of the steps and return the stats object. If not,
+        create a new mapping then return the stats object.
+
+        :param Host host: host to be determined.
+        """
+        if host.address not in self.host_stats:
+            self.host_stats[host.address] = HostStats
+        return self.host_stats[host.address]
+
+    def link_packet_loss(self, link, curr_time):
+        """
+        Record a packet loss of a link.
+
+        :param link: link of action.
+        :param curr_time: simulation time of loss.
+        """
+        stats = self.get_link_stats(link)
+        # TODO(team): Is this the most efficient?
+        stats.buffer_occupancy.append((curr_time, link.buffer_occupancy))
+        stats.packet_loss_times.append(curr_time)
+
+    def link_packet_transmitted(self, link, curr_time):
+        """
+        Record a packet transmission of a link.
+
+        :param link: link of action.
+        :param curr_time: simulation time of transmission.
+        """
+        stats = self.get_link_stats(link)
+        stats.buffer_occupancy.append((curr_time, link.buffer_occupancy))
+        stats.packet_transmit_times.append(curr_time)
+
+    def flow_packet_sent(self, flow, curr_time):
+        """
+        Record a packet sent from a flow.
+
+        :param flow: flow of action.
+        :param curr_time: simulation time of sending.
+        """
+        stats = self.get_flow_stats(flow)
+        stats.packet_sent_times_sec.append(curr_time)
+
+    def flow_packet_received(self, flow, curr_time):
+        """
+        Record a packet received by a flow.
+
+        :param flow: flow of action.
+        :param curr_time: simulation time of reception.
+        """
+        stats = self.get_flow_stats(flow)
+        stats.packet_rec_times_sec.append(curr_time)
+        # Retrieve the last element on the sent_times list to find
+        # the corresponding sent time and using the diff we can
+        # determine the rtt time.
+        sent_time = stats.packet_sent_times_sec[-1]
+        stats.packet_rtts_sec.append(curr_time - sent_time)
+
+    def host_packet_sent(self, host, curr_time):
```

Comment:
packet_sent_times is a list of (timestamp, packet_size) tuples. So you need to pass in the packet size as a variable in this function, and include that in the packet_sent_times list when you append.